components/zotero-protocol-handler.js

```diff
@@ -1230,9 +1225,8 @@ AsyncChannel.prototype = {
 					Zotero.debug("AsyncChannel: Got URI from generator");
 				}
 				
-				let uri = data;
-				uri.QueryInterface(Ci.nsIURL);
-				this.contentType = Zotero.MIME.getMIMETypeFromExtension(uri.fileExtension);
+				let sample = yield Zotero.File.getSample(data);
+				this.contentType = Zotero.MIME.getMIMETypeFromData(sample);
```

Comment:
If there's an extension, try `getMIMETypeFromExtension` first, to avoid reading the file unnecessarily?